source/jcli/cli.d

```diff
@@ -2,327 +2,1929 @@ module jcli.cli;
+
+        bool maybeMatchNextCommandNameAndResetState(string nameSlice)
+        {
+            bool didMatchCommand;
+
+            // This one HAS to stay inlined. Otherwise you get frame issues.
+            // See the older code: https://github.com/BradleyChatha/jcli/blob/511a02fe8dcd2913333f463ab5ad60d56fdb7f89/source/jcli/cli.d#L371-L430
+            // TODO: Add better match logic here.
+            matchSwitch: switch (nameSlice)
+            {
+                default:
+                {
+                    didMatchCommand = false;
+                    break matchSwitch;
+                }
+                static foreach (childNodeIndex, childNode; Graph.Adjacencies[typeIndex])
+                {{
+                    alias Type = CommandTypeContext.Types[childNode.typeIndex];
+                    static foreach (possibleName; CommandInfo!Type.udaValue.pattern)
+                    {
+                        case possibleName:
+                        {
+                            auto newCommand = commandIndexer!(childNode.typeIndex)(context).add();
+                            
+                            static if (childNode.fieldIndex != -1)
+                                newCommand.tupleof[childNode.fieldIndex] = command;
+                            
+                            context._matchedName = possibleName;
+                            didMatchCommand = true;
+                            break matchSwitch;
+                        }
+                    }
+                    
+                }}
+            }
+
+            if (didMatchCommand)
+            {
+                maybeReportParseErrorsFromFinalContext!ArgumentsInfo(parsingContext, errorHandler);
+                resetNamedArgumentArrayStorage!ArgumentsInfo(parsingContext);
+                tokenizer.resetWithRemainingRange();
+                context._state = State.matchedNextCommand;
+            }
+
+            return didMatchCommand;
+        }
+        
+        // Matched the given command.
+        while (!tokenizer.empty)
+        {
+            if (tryMatchSpecialThingsAndResetContextAccordingly(context, tokenizer))
+                return;
+
+            const currentToken = tokenizer.front;
+
+            auto consumeSingle()
+            { 
+                return consumeSingleArgumentIntoCommand!bindArgument(
+                    parsingContext, *command, tokenizer, errorHandler);
+            }
+
+            static if (0)
+            {
+                if (currentToken.kind.has(ArgToken.Kind.errorBit))
+                {
+                    // TODO: error handler integration, not sure yet??
+                    context._state = State.tokenizerError;
+                    // context._tokenizerError = currentToken.kind;
+                    return;
+                }
+            }
+
+            // We don't bother with subcommands if the current command has no children.
+            // Is this a hack? I'm not sure.
```

Comment:
@BradleyChatha I was referring more or less to the fact that I used the same function for both a single command execution and command discovery within modules, matching, and execution. But the fact that this piece of logic is hit even for the multiple-command case, I think justifies that it's alright.

I was debating whether faking that a root command has been matched in the single-command use case is a hack or not, but then decided it's ok precisely because that piece of logic is hit for both.